diff --git a/CMakeLists.txt b/CMakeLists.txt
index fb0ffa3..068dc46 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.5.1)
+cmake_minimum_required(VERSION 3.20)
 project(valijson)
 
 set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
@@ -29,26 +29,14 @@ add_library(valijson INTERFACE)
 # despite of valijson target is imported or not
 add_library(ValiJSON::valijson ALIAS valijson)
 
-include(GNUInstallDirs)
-target_include_directories(valijson INTERFACE
-        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
-        $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>)
+target_include_directories(valijson SYSTEM INTERFACE
+        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>)
+target_link_libraries(valijson INTERFACE nlohmann_json)
 
 if(valijson_USE_EXCEPTIONS)
     target_compile_definitions(valijson INTERFACE -DVALIJSON_USE_EXCEPTIONS=1)
 endif()
 
-install(DIRECTORY include/ DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
-
-install(TARGETS valijson
-    EXPORT valijsonConfig
-    DESTINATION ${CMAKE_INSTALL_LIBDIR}
-)
-
-install(EXPORT valijsonConfig
-    DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/valijson"
-)
-
 if(NOT valijson_BUILD_TESTS AND NOT valijson_BUILD_EXAMPLES)
     return()
 endif()
